crates/astria-sequencer-relayer/src/relayer.rs

```diff
@@ -84,75 +98,244 @@ impl Relayer {
-            data_availability_client,
+            sequencer,
             sequencer_poll_period: Duration::from_millis(cfg.block_time),
+            data_availability,
             validator,
-            block_tx,
+            gossip_block_tx,
             state_tx,
+            queued_blocks: Vec::new(),
+            conversion_workers: task::JoinSet::new(),
+            submission_task: None,
+            sequencer_task: None,
         })
     }
 
-    pub(crate) fn subscribe_to_state(&self) -> Receiver<State> {
+    pub(crate) fn subscribe_to_state(&self) -> watch::Receiver<State> {
         self.state_tx.subscribe()
     }
 
-    async fn get_and_submit_latest_block(&self) -> eyre::Result<State> {
-        let mut new_state = (*self.state_tx.borrow()).clone();
-        let resp = self.sequencer_client.get_latest_block().await?;
-
-        let height = resp.block.header.height.value();
-        if height <= *new_state.current_sequencer_height.get_or_insert(height) {
-            return Ok(new_state);
+    #[instrument(skip_all)]
+    fn handle_sequencer_tick(&mut self) {
+        use tendermint_rpc::Client as _;
+        if self.sequencer_task.is_none() {
+            let client = self.sequencer.clone();
+            self.sequencer_task = Some(tokio::spawn(async move { client.latest_block().await }));
+        } else {
+            debug!("task polling sequencer is currently in flight; not scheduling a new task");
         }
+    }
 
-        info!(height = ?height, tx_count = resp.block.data.len(), "got block from sequencer");
-        new_state.current_sequencer_height.replace(height);
+    #[instrument(skip_all)]
+    fn handle_sequencer_response(
+        &mut self,
+        join_result: Result<Result<block::Response, tendermint_rpc::Error>, task::JoinError>,
+    ) {
+        debug!("unsetting sequencer task");
+        self.sequencer_task = None;
+        // First check if the join task panicked
+        let request_result = match join_result {
+            Ok(request_result) => request_result,
+            // Report if the task failed, i.e. panicked
+            Err(e) => {
+                // TODO: inject the correct tracing span
+                report_err!(e, "sequencer poll task failed");
+                return;
+            }
+        };
+        match request_result {
+            Ok(block) => {
+                let current_height = self.state_tx.borrow().current_sequencer_height;
+                let validator = self.validator.clone();
+                // Start the costly conversion; note that the current height at
+                // the time of receipt matters. The internal state might have advanced
+                // past the height recorded in the block while it was converting, but
+                // that's ok.
+                self.conversion_workers.spawn_blocking(move || {
+                    convert_block_response_to_sequencer_block_data(block, current_height, validator)
+                });
```

Comment:
why is this costly?

Reply:
You have theoretically unbounded computation inside this. In a loop you: decode protobuf, (re)construct ed25519 keypairs, hash your data (including serialization), verify signatures, and do a few more things. So in that sense this conversion is CPU bound compared to most of the other execution here which is i/o bound.

The rule of thumb is that around 10ms in terms of waiting between await points.

This is more educated guessing then precise science. I guess it would be good to make actual benchmarks for the conversion method.